### PR TITLE
1400: Add Strive positive malaria cases bar graph + some core work

### PR DESCRIPTION
**Pull request by Alastair-L**
### Issue #: https://github.com/beyondessential/tupaia-backlog/issues/1400

To add the indicators required for the bar graph, one change was required in `data-broker > DhisService`.
### Situation:
Fetching event analytics with the new api.
2 programs: `SQMAL` and `SPF`. 
2 dataElementCodes, one from each program (`STR_QMAL05` and `STR_PF05`, respectively)
### Existing behavior:
The existing system failed when trying to fetch dataElementCodes that weren't part of the program being fetched, and tried to fetch both dataElements for both api calls.
### Proposed solution (this PR):
Only ever fetch dataElementCodes that are associated with the program being fetched, as according to the `data_source_data_group` table. 

Note: currently not all dataElement-program combinations are represented in that table, but that should be fixed here: https://github.com/beyondessential/tupaia/pull/1995

### I would love to hear feedback on whether this is a viable solution, or if there is another option we should pursue

### CODE REVIEW

File: packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js
Comment (Alastair-L) [inline packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js:307]: I just realized that this change might suppress some errors. Just a note to self to revert the change

File: packages/indicators/src/builders/helpers.ts
Comment (Alastair-L) [inline packages/indicators/src/builders/helpers.ts:45]: Will this have any unintended side affects?

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (Alastair-L) [inline packages/data-broker/src/services/dhis/DhisService.js:244]: Will this have any unintended side affects?

File: packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js
Comment (Alastair-L) [inline packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js:244]: Note to self: also could potentially hide errors?

File: packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js
Comment (Alastair-L) [reply packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js:307]: Actually does this make sense to consider as a case?

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (Alastair-L) [inline packages/data-broker/src/services/dhis/DhisService.js:200]: Do all these little, specific functions it easier to understand or less so?

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (Alastair-L) [inline packages/data-broker/src/services/dhis/DhisService.js:257]: Is stopping the fetching process if there are no `dataSources` to fetch going to have any unintended side affects?

File: packages/database/src/migrations/20210128055252-AddStrivePositiveResultBarGraph-modifies-data.js
Comment (Alastair-L) [inline packages/database/src/migrations/20210128055252-AddStrivePositiveResultBarGraph-modifies-data.js:43]: Note: because of the `SUM_EACH_WEEK` aggregationType for each dataElement in the indicators, there should only be one analytic per indicator per week returned to the dataBuilder.

This means that the entityAggregation here is mostly for show, the org unit should NOT be important in building the data. 

However, I figure it makes the most sense to still have it, as it matches conceptually what we are doing here.

File: packages/indicators/src/builders/helpers.ts
Comment (kael89) [reply packages/indicators/src/builders/helpers.ts:45]: I believe no! Good decision to use this option

File: packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js
Comment (kael89) [reply packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js:307]: I believe so! Example use case: get a list of programs, filter based on a condition that depends on dynamic data, then fetch using the remaining programs. The result could sometimes be an empty program list.

My conclusions so far:
* It is generally beneficial to support empty lists/maps as input
  ✔ supports a more flexible/dynamic usage of the API
  ✔enhances robustness
  ✔reduces complexity by introducing less assumptions
* On the other hand, we shouldn't overuse default parameters (`programCodes = []`) unless it is for really optional params eg optional config. In other words, support empty inputs, but the client of the API is expected to explicitly provide them.
  ✔ simpler function signatures
  ✔ no need to worry about `null` values
  ✔ we essentially assert the type of the param

File: packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js
Comment (kael89) [inline packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js:408]: Nice api usage!

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (kael89) [reply packages/data-broker/src/services/dhis/DhisService.js:200]: I love them! Size does matter in function length: all other qualitative aspects being equal, the smaller the better!

A few resources on the subject:
* [Martin Fowler - FunctionLength](https://martinfowler.com/bliki/FunctionLength.html)
* [Rule of 30](https://dzone.com/articles/rule-30-%E2%80%93-when-method-class-or)

**Review by kael89 — APPROVED** [review]
This is an excellent piece of work @Alastair-L !! Very mature work and approach on refactoring. I loved this PR 🎉 

Approving this since all my suggestions are about small details, but let's make sure that a few questions that arose are answered before we proceed

File: packages/database/src/migrations/20210128055252-AddStrivePositiveResultBarGraph-modifies-data.js
Comment (edmofro) [reply packages/database/src/migrations/20210128055252-AddStrivePositiveResultBarGraph-modifies-data.js:20]: I prefer 2 also:
- It makes the actual acronyms more obvious
- Just a bit friendlier on the eyes

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (nick-beyondessential) [reply packages/data-broker/src/services/dhis/DhisService.js:6]: @kael89 started looking at this with @IgorNadj, I suspect the `Event analytics require at least one organisation unit id` error doesn't get triggered because aggregator catches it first?
```
  async fetchAnalytics(codeInput, fetchOptions, aggregationOptions = {}) {
    const { organisationUnitCode, organisationUnitCodes } = fetchOptions;
    if (!organisationUnitCode && (!organisationUnitCodes || !organisationUnitCodes.length)) {
      // No organisation unit code, return empty response
      return {
        results: [],
        metadata: {
          dataElementCodeToName: {},
        },
        period: periodFromAnalytics([], fetchOptions),
      };
    }
```

Comment (kael89) [reply packages/data-broker/src/services/dhis/DhisService.js:6]: Yes, that makes sense! So an error will not get thrown, but the returned data will be always empty.

The next step would be to investigate existing functionality using `useDeprecatedApi: false` and confirm that it's broken - if it's not broken, then explain how it could work without org unit codes

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (Alastair-L) [reply packages/data-broker/src/services/dhis/DhisService.js:230]: Wow! Interesting. In fact it looks like all non-iterables don't error! Thanks
image
image

File: packages/database/src/migrations/20210128055252-AddStrivePositiveResultBarGraph-modifies-data.js
Comment (Alastair-L) [reply packages/database/src/migrations/20210128055252-AddStrivePositiveResultBarGraph-modifies-data.js:27]: Clean api 🎉

File: packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js
Comment (Alastair-L) [reply packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js:307]: Fantastic! That makes a lot of sense.

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (kael89) [reply packages/data-broker/src/services/dhis/DhisService.js:230]: I don't know this applied to other non-iterables, nice find!

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (Alastair-L) [reply packages/data-broker/src/services/dhis/DhisService.js:197]: Awesome! I finally found time to read those sources and they seem pretty clear! I'm glad to hear that (although perhaps they should be even shorter (`mergeAnalyticResults`, `mergeAnalyticsMetadata`)

**Review by kael89 — APPROVED** [review]

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (Alastair-L) [reply packages/data-broker/src/services/dhis/DhisService.js:239]: See [Nick's notes](https://github.com/beyondessential/tupaia/pull/1983#discussion_r575935785) on why it currently works and [Igor's addendum](https://github.com/beyondessential/tupaia/pull/2024#discussion_r576462039) on what the plan is, resulting in [this card](https://github.com/beyondessential/tupaia-backlog/issues/2251) to investigate further. Currently it looks like that card is a prerequisite for this one. Is that correct @IgorNadj?

File: packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js
Comment (Alastair-L) [reply packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js:244]: Taking a what would Kostas do approach, let's do what makes the code most closely match our intuitive separated concerns. Namely, that regardless of dataSource, you should only be able to pull for programCodes which are in the database. Making the logic part of the dhisApi doesn't make much sense as I'd expect that we want this functionality for all dataSources. So option 2 🎉

Comment (Alastair-L) [reply packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js]: Ended up only doing it in DhisService, as programCodes are only used in DHIS (which I didn't know)

**Comment by Alastair-L** [comment]
I believe that this PR is now just blocked on this: https://github.com/beyondessential/tupaia/pull/1983#discussion_r578143942

File: packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js
Comment (kael89) [reply packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js]: I guess after https://github.com/beyondessential/tupaia-backlog/issues/663 was implemented all elements are grouped under a data group, so we could use data groups/elements (and not surveys/questions)  for the `TupaiaService` too.

But we currently don't, so good call!

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (kael89) [inline packages/data-broker/src/services/dhis/DhisService.js:227]: It would be more efficient to fetch data sources using all `programCodes` with a single `find` query, then check whether any one is missing in the results

File: packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js
Comment (kael89) [inline packages/data-broker/src/__tests__/services/dhis/DhisService/testPull/testPullAnalytics.js:216]: (No need to change this) you can also use our custom matcher:

```suggestion
        ).toBeRejectedWith(/.*CODE_NOT_IN_SYSTEM.*/g);
```

Those are defined in [/jest.setup.js](https://github.com/beyondessential/tupaia/blob/dev/jest.setup.js)

**Review by kael89 — APPROVED** [review]
Just a small change suggested - making this optional as the performance gain for 1-2 programs would be negligible

**Comment by Alastair-L** [comment]
^ Small merge conflict with https://github.com/beyondessential/tupaia/pull/2124

**Review by kael89 — APPROVED** [review]

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (IgorNadj) [inline packages/data-broker/src/services/dhis/DhisService.js:138]: This commit would be a great separate PR. It would also involve regression testing a whole bunch of visualisations right?

**Comment by PShoe** [comment]
@Alastair-L or @IgorNadj checking on this PR. There's quite a bit to read in here and chasing where we landed. Still blocked on this [comment](https://github.com/beyondessential/tupaia/pull/1983#discussion_r578143942) ?

**Comment by kael89** [comment]
@PShoe without being 100% sure, my last impression about this card was that it was ready to go. Then there was this change: https://github.com/beyondessential/tupaia/pull/1983#discussion_r591910337, which as @IgorNadj commented may involve a significant amount of regression testing + it may not be required in the context of this PR.

**Note:** we need to take into account whether the changes here will work after https://github.com/beyondessential/tupaia-backlog/issues/2177 is merged. 2177 could actually make all the dhis-related changes in this PR redundant, although we can still keep them for future usage.

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (PShoe) [reply packages/data-broker/src/services/dhis/DhisService.js:217]: @IgorNadj  how does your investigation work affect the above?

**Comment by IgorNadj** [comment]
Discussed today, waiting on Alastair to have time to look into this with us

**Comment by Alastair-L** [comment]
# Notes:
## In https://github.com/beyondessential/tupaia/pull/1983#discussion_r579956717
>I guess after beyondessential/tupaia-backlog#663 was implemented all elements are grouped under a data group, so we could use data groups/elements (and not surveys/questions) for the TupaiaService too.

The discussed card is now implemented so we can clean up if desired.
## Response to https://github.com/beyondessential/tupaia/pull/1983#discussion_r591910337 
>This commit would be a great separate PR. It would also involve regression testing a whole bunch of visualisations right?

According to your notes here: https://github.com/beyondessential/tupaia-backlog/issues/2251#issuecomment-783090514, you're totally right this would break a preaggregator (plus a bunch of tests) (which I didn't test at the time 😬). Definitely worth reverting that change, not sure what was going through my head at that stage...

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (IgorNadj) [inline packages/data-broker/src/services/dhis/DhisService.js:162]: We used to merge first then translate, so we had to update the width and height to be correct (the translator relies on it - probs). After these changes, we translate first then merge, so we don't need this logic anymore

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (IgorNadj) [inline packages/data-broker/src/services/dhis/DhisService.js:168]: This method had been inlined, because the merging has been extracted into another method

File: packages/database/src/migrations/20210128055252-AddStrivePositiveResultBarGraph-modifies-data.js
Comment (IgorNadj) [inline packages/database/src/migrations/20210128055252-AddStrivePositiveResultBarGraph-modifies-data.js:45]: Data Broker does something funny (in pullAnalyticsFromEventsForApi). It lets you call getAnalytics with multiple program codes. 

BEFORE: It used to be that it would just loop the program codes and for each program request all the DE codes.  

AFTER: After these changes, it loops the program codes and for each program it only requests the DE codes for that program.

-----

Even though pullAnalyticsFromEventsForApi lets you do this, no vis/data builder does this, or it would fall over. TODO: check if this anyone actually does this (look at mapOverlay FRIDGE_SOH_VALUE). ------ And then check with Kostas to see if we can just throw an exception if any data builder tries to request data from dhis getAnalytics with >1 programCode, >1 DE.  

`programCodes` -> `programCode`.
`if (programCodes typeof Array) throw 'no'`

File: packages/data-broker/src/services/dhis/DhisService.js
Comment (rohan-bes) [reply packages/data-broker/src/services/dhis/DhisService.js:255]: This issue has been resolved a while ago, see: https://github.com/beyondessential/tupaia-backlog/issues/2251

**Comment by rohan-bes** [comment]
Closing out in favour of: https://github.com/beyondessential/tupaia/pull/2624

**Pull request closed without merging**
